components/script/dom/worklet.rs

```diff
@@ -0,0 +1,621 @@
+                    self.gc();
+                } else if !self.should_gc {
+                    self.should_gc = true;
+                    let msg = WorkletData::StartSwapRoles(self.role.sender.clone());
+                    let _ = self.hot_backup_sender.send(msg);
+                }
+            }
+        }
+    }
+
+    /// The current memory usage of the thread
+    #[allow(unsafe_code)]
+    fn current_memory_usage(&self) -> u32 {
+        unsafe { JS_GetGCParameter(self.runtime.rt(), JSGCParamKey::JSGC_BYTES) }
+    }
+
+    /// Perform a GC.
+    #[allow(unsafe_code)]
+    fn gc(&mut self) {
+        debug!("BEGIN GC (usage = {}, threshold = {}).", self.current_memory_usage(), self.gc_threshold);
+        unsafe { JS_GC(self.runtime.rt()) };
+        self.gc_threshold = max(MIN_GC_THRESHOLD, self.current_memory_usage() * 2);
+        debug!("END GC (usage = {}, threshold = {}).", self.current_memory_usage(), self.gc_threshold);
+    }
+
+    /// Get the worklet global scope for a given worklet.
+    /// Creates the worklet global scope if it doesn't exist.
+    fn get_worklet_global_scope(&mut self,
+                                pipeline_id: PipelineId,
+                                worklet_id: WorkletId,
+                                global_type: WorkletGlobalScopeType,
+                                base_url: ServoUrl)
+                                -> Root<WorkletGlobalScope>
+    {
+        match self.global_scopes.entry(worklet_id) {
+            hash_map::Entry::Occupied(entry) => Root::from_ref(entry.get()),
+            hash_map::Entry::Vacant(entry) => {
+                let result = global_type.new(&self.runtime, pipeline_id, base_url, &self.global_init);
+                entry.insert(JS::from_ref(&*result));
+                result
+            },
+        }
+    }
+
+    /// Fetch and invoke a worklet script.
+    ///g https://drafts.css-houdini.org/worklets/#fetch-and-invoke-a-worklet-script
+    fn fetch_and_invoke_a_worklet_script(&self,
+                                         global_scope: &WorkletGlobalScope,
+                                         script_url: ServoUrl,
+                                         credentials: RequestCredentials,
+                                         pending_tasks_struct: PendingTasksStruct,
+                                         promise: TrustedPromise)
+    {
+        debug!("Fetching from {}.", script_url);
+        // Step 1.
+        // TODO: Settings object?
+
+        // Step 2.
+        // TODO: Fetch a module graph, not just a single script.
+        // TODO: Fetch the script asynchronously?
+        // TODO: Caching.
+        let resource_fetcher = self.global_init.resource_threads.sender();
+        let request = RequestInit {
+            url: script_url,
+            type_: RequestType::Script,
+            destination: Destination::Script,
+            credentials_mode: credentials.into(),
+            .. RequestInit::default()
```

Comment:
CORS mode isn't a problem, but I'm not sure which origin to use here. One snag is that this is being done in a different thread, so we don't have access to the `MutableOrigin` of the window.